blog/modules/ROOT/pages/using-types.adoc

```diff
@@ -0,0 +1,220 @@
+= Taking types to the next level
+
+:use-link-attrs:
+
+// tag::byline[]
+++++
+<div class="blog-byline">
+++++
+by link:https://github.com/stackoverflow[Islon Scherer] on October 1st, 2024
+++++
+</div>
+++++
+// end::byline[]
+
+// tag::excerpt[]
+One of the main points of using Pkl is to describe what your configuration looks like.
+An often overlooked design goal (arguably, the most important), is to
+*forbid* invalid configurations. Pkl types allow a great degree of freedom
+and power in constraining what your data looks like.
+// end::excerpt[]
+
+Some developers, especially those coming from more general purpose languages like
+Go or Java, may have a pre-defined notion of how types can help them.
+So let's have a look at how Pkl can go beyond most traditional types.
+
+== Beyond Strings
+
+`String` is a type we tend to see everywhere. This is especially true in configuration
+languages, where a `String` can be used as a "joker" type for many different values.
+
+=== Enumerating possibilities
+
+A typical usage of Strings is to define a finite set of values that your property can take:
+
+[source,pkl]
+----
+/// The environment where this service is running.
+///
+/// Valid possibilities are "dev", "qa" and "prod".
+env: String
+----
+
+This works, but has the problem that users of your template can provide invalid data, like `"foo"`.
+So let's do better:
+
+[source,pkl]
+----
+/// The environment where this service is running.
+env: "dev"|"qa"|"prod"
+----
+
+Now `env` will reject any invalid possibility. If you're using an IDE, like the
+link:https://github.com/apple/pkl-intellij[IntelliJ plugin], this has the added benefits
+of providing autocomplete and telling you when a given value is invalid.
+
+In Pkl, strings are not only values, they can also be types.
+
+=== Matching values
+
+Another typical use for strings is to allow values that have a specific *shape*:
+
+[source,pkl]
+----
+/// The IPv4 to connect
+serverIp: String
+----
+
+Again, we can do better to validate that the given ip is correct:
+
+[source,pkl]
+----
+/// The IPv4 to connect
+serverIp: String(matches(Regex(#"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}"#))) // <1>
+----
+<1> This regex is kept simple for illustration purposes, it still allows for invalid inputs.
```

Comment:
Can use `((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}` instead, which is even shorter and seems to be pretty good. See https://stackoverflow.com/a/36760050/2460186